protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,411 @@
+                    }
+                    relayed_multiaddr.dst_peer_id = Some(peer_id)
+                }
+            }
+            p => {
+                if before_circuit {
+                    relayed_multiaddr
+                        .relay_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                } else {
+                    relayed_multiaddr
+                        .dst_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                }
+            }
+        }
+    }
+
+    Ok(Ok(relayed_multiaddr))
+}
+
+pub struct RelayListener {
+    queued_new_addresses: VecDeque<Multiaddr>,
+    from_behaviour: mpsc::Receiver<ToListenerMsg>,
+    msg_to_behaviour: Option<BoxFuture<'static, Result<(), mpsc::SendError>>>,
+}
+
+impl Unpin for RelayListener {}
+
+impl Stream for RelayListener {
+    type Item =
+        Result<ListenerEvent<Ready<Result<RelayedConnection, RelayError>>, RelayError>, RelayError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(msg) = &mut self.msg_to_behaviour {
+                match Future::poll(msg.as_mut(), cx) {
+                    Poll::Ready(Ok(())) => self.msg_to_behaviour = None,
+                    Poll::Ready(Err(e)) => return Poll::Ready(Some(Err(e.into()))),
+                    Poll::Pending => {}
+                }
+            }
+
+            if let Some(addr) = self.queued_new_addresses.pop_front() {
+                return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr))));
+            }
+
+            match self.from_behaviour.poll_next_unpin(cx) {
+                Poll::Ready(Some(ToListenerMsg::IncomingRelayedConnection {
+                    stream,
+                    src_peer_id,
+                    relay_addr,
+                    relay_peer_id: _,
+                })) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                        upgrade: ready(Ok(stream)),
+                        local_addr: relay_addr.with(Protocol::P2pCircuit),
+                        remote_addr: Protocol::P2p(src_peer_id.into()).into(),
+                    })));
+                }
+                Poll::Ready(Some(ToListenerMsg::Reservation(Ok(Reservation { addrs })))) => {
+                    debug_assert!(
+                        self.queued_new_addresses.is_empty(),
+                        "Assert empty due to previous `pop_front` attempt."
+                    );
+                    // Returned as [`ListenerEvent::NewAddress`] in next iteration of loop.
+                    self.queued_new_addresses = addrs.into();
+                }
+                Poll::Ready(Some(ToListenerMsg::Reservation(Err(())))) => {
+                    return Poll::Ready(Some(Err(RelayError::Reservation)));
+                }
+                Poll::Ready(None) => {
```

Comment:
Same as in the handler, I would find it easier to read if we decompose the value in steps and panic early if it is `None`.

Reply:
`panic!` removed with 4b90626ab232faacba28139471f7c6fba0b520ab. I guess we could do `ready!(self.from_behaviour.poll_next_unpin(cx))?`, but I think I prefer the explicit return on `None`.

@thomas I am not sure I follow with your above suggestion of "decomposing the value".

Reply:
Off-topic: Make sure to tag the right thomas (@thomaseizinger !) :)

> @thomas I am not sure I follow with your above suggestion of "decomposing the value".

Like this: https://github.com/thomaseizinger/rust-libp2p/commit/fc73f1459377936a85e3f9baf822fe9d2f2f6839

I'd argue that the logic is easier to understand if the control flow is separated as much as possible and explicit. This patch uses `match` always as an expression and only uses control flow keywords (`return` and `continue`) to break _out_ of the top-to-bottom happy-path flow.

Reply:
Sounds good to me. Cherry picked as 8aa386739dbb73d5f59461acc7cc8e6945b39f91. 

> This patch uses match always as an expression and only uses control flow keywords (return and continue) to break out of the top-to-bottom happy-path flow.

Makes sense. I will try to keep this in mind when writing these going forward. Please do remind me.